### PR TITLE
[enhancement](nereids)allow reorder mark join

### PR DESCRIPTION
**Pull request by starocean999**
## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinCommute.java [hunk 1]
```diff
@@ -59,7 +60,8 @@ public Rule build() {
                 .when(join -> check(swapType, join))
                 .whenNot(LogicalJoin::hasDistributeHint)
                 .whenNot(join -> joinOrderMatchBitmapRuntimeFilterOrder(join))
-                .whenNot(LogicalJoin::isMarkJoin)
+                .whenNot(join -> JoinUtils.isNullAwareMarkJoin(join))
+                .whenNot(join -> join.isMarkJoin() && join.getHashJoinConjuncts().isEmpty())
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/join/JoinCommute.java:64]: why cannot commute? add some comment?

**Comment by starocean999** [comment]
run buildall

**Comment by starocean999** [comment]
run buildall

File: regression-test/data/nereids_tpcds_shape_sf1000_p0/shape/query10.out [hunk 2]
```diff
@@ -10,48 +10,48 @@ PhysicalResultSink
...
---------------------hashJoin[LEFT_SEMI_JOIN] hashCondition=((c.c_customer_sk = catalog_sales.cs_ship_customer_sk)) otherCondition=()
-----------------------hashJoin[LEFT_SEMI_JOIN] hashCondition=((c.c_customer_sk = web_sales.ws_bill_customer_sk)) otherCondition=()
-------------------------PhysicalDistribute[DistributionSpecHash]
---------------------------PhysicalProject
-----------------------------hashJoin[INNER_JOIN] hashCondition=((customer_demographics.cd_demo_sk = c.c_current_cdemo_sk)) otherCondition=() build RFs:RF5 c_current_cdemo_sk->[cd_demo_sk]
-------------------------------PhysicalOlapScan[customer_demographics] apply RFs: RF5
-------------------------------PhysicalDistribute[DistributionSpecReplicated]
---------------------------------PhysicalProject
-----------------------------------hashJoin[RIGHT_SEMI_JOIN] hashCondition=((c.c_customer_sk = store_sales.ss_customer_sk)) otherCondition=() build RFs:RF4 c_customer_sk->[ss_customer_sk]
+--------------------hashJoin[RIGHT_SEMI_JOIN] hashCondition=((c.c_customer_sk = catalog_sales.cs_ship_customer_sk)) otherCondition=()
```
Comment (xzj7019) [inline regression-test/data/nereids_tpcds_shape_sf1000_p0/shape/query10.out:13]: real benefit would better verified on cluster.

**Review by xzj7019 — APPROVED** [review]

**Review by morrySnow — APPROVED** [review]

**Pull request merged**